Add support for IoT class in manifest

frenck opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

It recently came up on Discord again, and I still had the code around. This suggests on adding the IoT class to the manifest (as a central place for both the UI & Docs).

During Hacktoberfest of 2020, we had all IoT class data completed on the documentation, so that makes this quite a nice migration set.

The migration is done in the last commit. This PR is best reviewed by checking the first commits.

Dev docs PR: <https://github.com/home-assistant/developers.home-assistant/pull/886>

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

frenck committed:
Add support for IoT class in manifest

bdraco reviewed (approved):
LGTM.

Are these ever going to need to be translated?

frenck commented:
Translations for documentation: I don't think so, as well kinda decided upon that that won't be happening. As for the backend/frontend use, I guess so.

bdraco commented:
> Translations for documentation: I don't think so, as well kinda decided upon that that won't be happening. As for the backend/frontend use, I guess so. Why you ask? 😀 

If they were going to end up in a translations file as keys it might have made sense to have them be snake case.

balloob commented on script/hassfest/manifest.py:
```diff
@@ -15,6 +15,14 @@
 SUPPORTED_QUALITY_SCALES = ["gold", "internal", "platinum", "silver"]
+SUPPORTED_IOT_CLASSES = [
+    "Assumed State",
+    "Calculated",
```
I think that we shouldn't have Calculated for "internal" integrations. I think that they should just have none.

frenck replied:
I'm not sure what that means?
Calculated means: 

> The integration does not handle communication on it's own, but calculates the state.

Let's consider `sun`... its internal, but its state is clearly calculated. Or the `updater` integration, is internal, but is cloud polling.

I'm not sure why we would hide that?

balloob replied:
Yeah, the `sun` is fine. So is `updater`. But setting `default_config` to calculated seems wrong. I would even argue that `counter` should not be `calculated` because it's user input.

frenck commented:
> If they were going to end up in a translations file as keys it might have made sense to have them be snake case.

Agree, will adjust that 👍

frenck commented:
OK, after some discussion, we going for lower and snake case the IoT classes. Additionally, we keep with the current set of classes, but "Calculated" is going to be used limited for only things like "Sun". Input helpers, for example, will not have a IoT class.

frenck committed:
Adjust hassfest + strictness

frenck commented:
- Rebased the PR
- Adjusted IoT class types/list based on snake_case
- Added more strictness by defining a list that shouldn't have an IoT class, allowing to require it for others.
- Stripped of the commit that added all IoT classes.

The build will fail now. Will do separate commits to resolve the hassfest checks failures.

frenck committed:
Extend/correct no IoT Class list

frenck committed:
Migrate IoT Classes

bdraco reviewed (approved):
Nice!

frenck commented:
Awaiting merge in result of this comment:

<https://github.com/home-assistant/developers.home-assistant/pull/886#pullrequestreview-635980597>

There might be a misunderstanding that may result in changing the code in this PR.
